common/customizable_schema.go

```diff
@@ -13,9 +15,43 @@ type CustomizableSchema struct {
 	Schema         *schema.Schema
 	path           []string
 	isSuppressDiff bool
+	context        SchemaPathContext
 }
 
-func CustomizeSchemaPath(s map[string]*schema.Schema, path ...string) *CustomizableSchema {
+// This context is used in CustomizeSchemaPath function to proivde information about nested schema referenced by the schema registry.
+type SchemaPathContext struct {
+	// Path is used for refernces from resourceProviderRegistry as prefix, follows the format of `[a,0,b]`
+	path       []string
+	schemaPath []*schema.Schema
+}
+
+func (spc SchemaPathContext) copy() SchemaPathContext {
+	newPath := make([]string, len(spc.path))
+	copy(newPath, spc.path)
+	newSchemaPath := make([]*schema.Schema, len(spc.schemaPath))
+	copy(newSchemaPath, spc.schemaPath)
+	return SchemaPathContext{
+		path:       newPath,
+		schemaPath: newSchemaPath,
+	}
+}
+
+func (spc SchemaPathContext) addToPath(fieldName string, schema *schema.Schema) SchemaPathContext {
+	newSpc := spc.copy()
+	// Special path element `"0"` is used to denote either arrays or sets of elements
+	newSpc.path = append(spc.path, fieldName, "0")
+	newSpc.schemaPath = append(spc.schemaPath, schema)
+	return newSpc
+}
+
+func getEmptySchemaPathContext() SchemaPathContext {
+	return SchemaPathContext{
+		[]string{},
+		[]*schema.Schema{},
+	}
+}
+
+func CustomizeSchemaPath(ctx SchemaPathContext, s map[string]*schema.Schema, path ...string) *CustomizableSchema {
```

Comment:
I feel that this is a less usable interface and requires more boilerplate for people to use. What design spaces can we consider here?

Personally, I think it would make a lot of sense for SchemaPathContext to include the schema itself. That way you only have one parameter to work with.

Another variation would be to include this information in `CustomizableSchema` and construct this within `reflect_resource.go` and pass it to users. We would need some methods to index into this structure then. Then a user could do something like:

```go
func (r Resource) CustomizeSchema(s *CustomizableSchema) *CustomizableSchema {
  s.SchemaPath("field1", "attribute").SetRequired()
  s.SchemaPath("field2").SetConflictsWith("field3")
  // It is now unsafe to modify the schema directly, but we can expose if needed.
  s.Underlying.ConflictsWith = []string{"field4"}
}
```
Then this is all entirely hidden from the user's perspective.

Reply:
Yeah good point, I do agree that exposing a new argument is not very easy to use. I created an updated version here https://github.com/databricks/terraform-provider-databricks/pull/3481, can you take another look? Thanks!